lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,464 @@
+        !Bool
+          -- ^ Is key 0 still registered? For compatibility with
+          -- single-stake-key wallets, we need to track this.
+          --
+          -- See the implementaiton of @applyTx@ for how it is used.
+    deriving (Eq, Show, Generic)
+
+instance NFData State
+
+instance (NFData (k 'AccountK XPub), NFData (k 'AddressK XPub))
+    => NFData (DelegationState k)
+
+deriving instance
+    ( Show (k 'AccountK XPub)
+    , Show (k 'AddressK XPub)
+    ) => Show (DelegationState k)
+
+deriving instance
+    ( Eq (k 'AccountK XPub)
+    , Eq (k 'AddressK XPub)
+    ) => Eq (DelegationState k)
+
+pointer :: DelegationState k -> Maybe PointerUTxO
+pointer (DelegationState _ (More _ p _)) = Just p
+pointer _ = Nothing
+
+keyAtIx
+    :: (SoftDerivation k)
+    => DelegationState k
+    -> Index 'Soft 'AddressK
+    -> k 'AddressK XPub
+keyAtIx s = deriveAddressPublicKey (rewardAccountKey s) MutableAccount
+
+nextKeyIx
+    :: DelegationState k
+    -> Index 'Soft 'AddressK
+nextKeyIx s = case state s of
+    Zero -> minBound
+    One -> succ minBound
+    More ix _ _ -> ix
+
+lastActiveIx
+    :: DelegationState k
+    -> Maybe (Index 'Soft 'AddressK)
+lastActiveIx s
+    | nextKeyIx s == minBound = Nothing
+    | otherwise               = Just $ pred $ nextKeyIx s
+
+data PointerUTxO = PointerUTxO { pTxIn :: TxIn, pCoin :: Coin }
+    deriving (Generic, Eq, Show)
+    deriving anyclass NFData
+
+-- | Returns the index corresponding to the payment key the `PointerUTxO`
+-- should be locked with for a portfolio of a given size @n@.
+--
+-- In our current implementation we require the `PointerUTxO` to be created in
+-- the @[0] -> [0,1] transition@, i.e. @nextKeyIx 1 -> nextKeyIx 2@.
+pointerIx
+    :: Int
+    -> Maybe (Index 'Soft 'AddressK)
+pointerIx 0 = Nothing
+pointerIx 1 = Nothing
+pointerIx n = Just $ toEnum n
+
+--------------------------------------------------------------------------------
+-- Constructors
+--------------------------------------------------------------------------------
+
+-- | Construct the initial delegation state.
+initialDelegationState
+    :: k 'AccountK XPub
+    -> DelegationState k
+initialDelegationState accK = DelegationState accK Zero
+
+--------------------------------------------------------------------------------
+-- Chain
+--------------------------------------------------------------------------------
+
+-- | A transaction type specific to `DelegationState`.
+--
+-- Intented to be converted both from and to a more real transaction type.
+--
+-- When constructing a real transaction from `Tx`, these `outputs`
+-- should appear before other outputs. In the theoretical event that there's
+-- also a user-specified output with the same payment key as the pointer output,
+-- `applyTx` will track the first one as the new pointer.
+data Tx = Tx
+    { certs :: [Cert]
+    , inputs :: [(TxIn, Coin)]
+    , outputs :: [TxOut]
+    }
+    deriving (Eq, Generic)
+    deriving Show via (Quiet Tx)
```

Comment:
You could potentially use a type-class for that to be more like a DSL for which the primitive Tx is one possible implementation. No need to perform conversions back and forth with this; only an instance of the class is needed which is likely trivial to write. 

```hs
class DelegationStateTx tx of
    certs :: tx -> [Cert]
    inputs :: tx -> [(TxIn, Coin)]
    outputs :: tx -> [TxOut]
```

Reply:
I think it's neat to return the same `Tx` type from `cmdSetPortfolioOf`, as what we take in `applyCmd`. And I like defaulting to no type classes.

Otherwise you'd have some kind of `mkTx :: certs -> ins -> outs -> tx` (which might not make sense), or some  "`TxRecipe`" in addition.